cf-agent/verify_files.c

```diff
@@ -559,6 +581,73 @@ static PromiseResult VerifyFilePromise(EvalContext *ctx, char *path, const Promi
 
 /*****************************************************************************/
 
+static PromiseResult WriteContentFromString(EvalContext *ctx, const char *path, const Attributes *attr,
+                                            const Promise *pp)
+{
+    assert(path != NULL);
+    assert(attr != NULL);
+
+    PromiseResult result = PROMISE_RESULT_NOOP;
+
+    if (attr->content == NULL)
+    {
+        RecordFailure(ctx, pp, attr, "'content' not set for promiser: '%s'", path);
+        return PromiseResultUpdate(result, PROMISE_RESULT_FAIL);
+    }
+
+    unsigned char existing_content_digest[EVP_MAX_MD_SIZE + 1] = { 0 };
+    if (access(path, R_OK) == 0)
+    {
+        HashFile(path, existing_content_digest, CF_DEFAULT_DIGEST,
+                 FileNewLineMode(path) == NewLineMode_Native);
+    }
+
+    unsigned char promissed_content_digest[EVP_MAX_MD_SIZE + 1] = { 0 };
+    HashString(attr->content, strlen(attr->content),
+               promissed_content_digest, CF_DEFAULT_DIGEST);
+
+    if (!HashesMatch(existing_content_digest, promissed_content_digest, CF_DEFAULT_DIGEST))
+    {
+        if (DONTDO)
+        {
+            RecordWarning(ctx, pp, attr,
+                          "Should update content of '%s' with content '%s'",
+                          path, attr->content);
+            return result;
+        }
+
+        FILE *f = safe_fopen(path, "w");
+        if (f == NULL)
+        {
+            RecordFailure(ctx, pp, attr, "Cannot open file '%s' for writing", path);
+            return PromiseResultUpdate(result, PROMISE_RESULT_FAIL);
+        }
+
+        Writer *w = FileWriter(f);
+        size_t bytes_to_write = strlen(attr->content);
```

Comment:
I would move this up, before the other place you do `strlen(attr->content)`, and use the variable to save some cycles iterating over the same data multiple times.